### PR TITLE
refactor workflow package

### LINKED ISSUE #1743: Proposal: Move function SubmitWorkflow
The function [`SubmitWorkflow`](https://github.com/sql-machine-learning/sqlflow/blob/44939623ee243d06f5d7b9f3dac4e8c5d8b7c4b3/pkg/sql/executor_ir.go#L80) defined in package `pkg/sql` is only used in [`cmd/sqlflowserver`](https://github.com/sql-machine-learning/sqlflow/blob/44939623ee243d06f5d7b9f3dac4e8c5d8b7c4b3/cmd/sqlflowserver/main.go#L63) together with `pkg/server.NewServer`.

Hence, I propose to move the definition of this function to `pkg/server` as `submitWorkflow`.

A side-effect of this proposal is to minimize and finally remove the file `execute_ir.go`, because the IR is not executable -- it is the codegen result, a.k.a., submitter programs, which are executable.

### PR DESCRIPTION
**Pull request by Yancey0623**
part work of #1980 
fixed #1743

### CODE REVIEW

File: cmd/sqlflow/main.go [hunk 1]
```diff
@@ -42,7 +42,8 @@ func compile(cgName, sqlProgram, datasource string) (string, error) {
 		}
 		sess := sql.MakeSessionFromEnv()
 		sess.DbConnStr = datasource
-		return couler.GenCode(spIRs, sess)
+		cg := &couler.Codegen{}
```
Comment (typhoonzero) [inline cmd/sqlflow/main.go]: `cg` -> `codegen` maybe easy to read?

File: pkg/server/sqlflowserver.go [hunk 2]
```diff
@@ -112,3 +117,31 @@ func (s *Server) Run(req *pb.Request, stream pb.SQLFlow_RunServer) error {
 	}
 	return nil
 }
+
+// SubmitWorkflow submits an Argo workflow
+//
+// TODO(wangkuiyi): Make SubmitWorkflow return an error in addition to
+// *pipe.Reader, and remove the calls to log.Printf.
+func SubmitWorkflow(sqlProgram string, modelDir string, session *pb.Session) *pipe.Reader {
```
Comment (typhoonzero) [inline pkg/server/sqlflowserver.go:124]: Why not put this function under `pkg/workflow`? and rename to `Submit` so the caller side would use `workflow.Submit()`.

File: pkg/workflow/argo/submit.go [hunk 3]
```diff
@@ -14,6 +14,6 @@
 package argo
 
 // Submit the Argo workflow and returns the workflow ID
-func Submit(argoYAMLContent string) (string, error) {
+func (w Workflow) Submit(argoYAMLContent string) (string, error) {
```
Comment (weiguoz) [inline pkg/workflow/argo/submit.go]: pointer type?
`(w *Workflow) Submit()`

File: pkg/server/sqlflowserver.go (hunk 2 above)
Comment (Yancey0623) [reply pkg/server/sqlflowserver.go:124]: `SubmitWorkfow ` implements `run` in the `Server` struct, and maybe `workflow.Submit` returns a workflow ID is more meaningful than `*pip.Reader`

File: pkg/workflow/workflow.go [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+// New returns Codegen and Workflow instance
+func New(backend string) (Codegen, Workflow, error) {
+	if backend == "argo" {
+		return &couler.Codegen{}, &argo.Workflow{}, nil
+	}
+	return nil, nil, fmt.Errorf("the specifiy backend: %s has not support", backend)
+}
+
+// ResolveAndSubmitWorkflow resolve sql program to IRs and submit workflow YAML to Kubernetes
```
Comment (wangkuiyi) [inline pkg/workflow/workflow.go]: sql program => a SQL program

Comment (wangkuiyi) [inline pkg/workflow/workflow.go]: resolve => compiles

submit => submits

ResolveAndSubmitWorkflow => CompileAndSubmitWorkflow => workflow.Run

File: pkg/workflow/workflow.go [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+import (
+	"fmt"
+
+	"sqlflow.org/sqlflow/pkg/database"
+	"sqlflow.org/sqlflow/pkg/ir"
+	"sqlflow.org/sqlflow/pkg/parser"
+	pb "sqlflow.org/sqlflow/pkg/proto"
+	"sqlflow.org/sqlflow/pkg/sql"
+	"sqlflow.org/sqlflow/pkg/workflow/argo"
+	"sqlflow.org/sqlflow/pkg/workflow/codegen/couler"
```
Comment (wangkuiyi) [inline pkg/workflow/workflow.go]: Should we name the package `workflow/codegen/couler` or  `workflow/couler`?

File: pkg/workflow/workflow.go [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+
+// Workflow submits workflow task and trace step status
+type Workflow interface {
+	Submit(string) (string, error)
+	Fetch(*pb.FetchRequest) (*pb.FetchResponse, error)
+}
+
+// New returns Codegen and Workflow instance
+func New(backend string) (Codegen, Workflow, error) {
+	if backend == "argo" {
```
Comment (wangkuiyi) [inline pkg/workflow/workflow.go]: "argo" => "couler" ?

File: cmd/sqlflow/main.go (hunk 1 above)
Comment (wangkuiyi) [inline cmd/sqlflow/main.go]: Giving that `workflow.New("argo")` returns a `workflow/codegen/couler.Codegen`, I think we might want users to call `workflow.New("argo")`

File: pkg/workflow/workflow.go (hunk 5 above)
Comment (Yancey0623) [reply pkg/workflow/workflow.go]: It's preparing for future import `Fluid`, I would remove the `codegen` folder and re-add it if necessary in the future.

File: pkg/server/sqlflowserver.go [hunk 7]
```diff
@@ -112,3 +116,37 @@ func (s *Server) Run(req *pb.Request, stream pb.SQLFlow_RunServer) error {
...
+	logger := log.WithFields(log.Fields{
+		"requestID": log.UUID(),
+		"user":      session.UserId,
+		"submitter": session.Submitter,
+		"event":     "submitWorkflow",
+	})
+	if os.Getenv("SQLFLOW_WORKFLOW_LOGVIEW_ENDPOINT") == "" {
+		logger.Fatalf("should set SQLFLOW_WORKFLOW_LOGVIEW_ENDPOINT if enable argo mode.")
```
Comment (typhoonzero) [inline pkg/server/sqlflowserver.go:132]: if enable argo mode -> if Argo mode is enabled.

File: pkg/server/sqlflowserver.go [hunk 8]
```diff
@@ -112,3 +116,37 @@ func (s *Server) Run(req *pb.Request, stream pb.SQLFlow_RunServer) error {
...
+	}
+	rd, wr := pipe.Pipe()
+	startTime := time.Now()
+	go func() {
+		defer wr.Close()
+		wfID, e := workflow.Run("couler", sqlProgram, session, logger)
+		defer logger.Infof("submitted, workflowID:%s, spent:%.f, error:%v", wfID, time.Since(startTime).Seconds(), e)
```
Comment (typhoonzero) [inline pkg/server/sqlflowserver.go:139]: 1. add space after `:`
2. spent:%.f -> spent: %.f sec

**Review by typhoonzero — APPROVED** [review]
Just some tiny suggestions.

**Pull request merged**